### DISCUSSION TITLE
How to switch on/off charging/discharging in iobroker?

### DISCUSSION THREAD

**Discussion by rolandm1234** [syssi/esphome-jbd-bms#39] [Q&A] [answered]
I have flashed a esp32 with ble firmware, connected via mqtt to iobroker and get all sensors in iobroker.
i have the datapoints jbd-bms-ble/switch/jbd-bms-discharging/command and state but cannot switch discharging on/off.

what is the syntax?

**Comment by syssi** [comment]
Controls aren't implemented yet. I did add the switch entities already but the don't issue any commands to the BMS yet.

**Reply by syssi** [reply, marked as answer]
Please track the progress of the merge request: https://github.com/syssi/esphome-jbd-bms/pull/36

As soon it's merged you can control the switches using the command topic with the payload `ON` and `OFF`.

**Reply by syssi** [reply]
If you change this substitution from:

```
substitutions:
  external_components_source: github://syssi/esphome-jbd-bms@main
```

to 

```
substitutions:
  external_components_source: github://syssi/esphome-jbd-bms@add-controls
```

and flash your ESP again you could test the mosfet switches already!

**Reply by rolandm1234** [reply]
```
roland@ro02:~/esphome-jbd-bms$ esphome run esp32-ble-example.yaml 
INFO Reading configuration esp32-ble-example.yaml...
INFO Updating https://github.com/syssi/esphome-jbd-bms.git@add-controls
Failed config

switch.jbd_bms_ble: [source esp32-ble-example.yaml:211]
  platform: jbd_bms_ble
  jbd_bms_ble_id: bms0
  charging: 
    name: jbd-bms-ble charging
  discharging: 
    name: jbd-bms-ble discharging
  
  [balancer] is an invalid option for [switch.jbd_bms_ble]. Please check the indentation.
  balancer: 
    name: jbd-bms-ble balancer
```

**Reply by syssi** [reply]
Please remove the `balancer` switch from your YAML temporary.

**Reply by rolandm1234** [reply]
ok, switching charging/discharging is working now.

**Reply by moriss1919** [reply]
hello
I've tried this solution also. I thinked it was ok but not completely. When I switch off charging or discharging switch, it turn off but immediately it turn ON itself... Any idea ?

**Reply by syssi** [reply]
I have an idea what's going on.

**Reply by moriss1919** [reply]
ok cool :D

**Reply by moriss1919** [reply]
and the idea is ?

**Reply by syssi** [reply]
The implementation isn't perfect yet. Both Mosfet states (charging on/off, discharging on/off) are stored at the same byte (bitmask). To change one of the mosfet states this byte must be updated/written. The implementation must keep track of both states to be able to update the byte properly.

1. If you disable/enable the charging switch only (once per minute). Does it work without trouble?
2. If you disable/enable the discharging switch only (once per minute). Does it work without trouble?
3. Change the state of the charging switch + wait 30 seconds + change the states of the discharging switch. Does it work without trouble?
4. If you change the charging & discharging switch simultaneously the result will be unpredictable. Correct?
